[1/n][export] Refactor PT2 Archive weight saving and loading

yiming0416 opened the pull request:
Summary:

We split the refactoring in two parts for forward compatibility concerns
First, we land the deserialization (loading part)
Then, we land the serialization (saving part)

Save weights and constants as individual files in PT2 archive. Each weight/constant will be saved as raw bytes, unless it is a custom object (TorchBind object) or a non-fake tensor subclass, for these two special cases we still save them using pickle.

The metadata of saved tensors along with the file name will be saved as `PayloadMeta`.
The mapping from FQN to `PayloadMeta` will be saved as `PayloadConfig` under `WEIGHTS_CONFIG_FORMAT` and `CONTANTS_CONFIG_FORMAT`

This changes the serialization in python side when calling `torch.export.save()`.

For deserialization in python `torch.export.load()`, we make it BC-safe by allowing loading legacy format weights/constants.

For deserialization in C++ `torch/nativert/ModelRunner.cpp`, we make this a BC breaking change as currently the OSS ModelRunner API is not being used.

The file structure

```
├── archive_format
├── archive_version
├── byteorder
├── .data
│   ├── serialization_id
│   └── version
├── data
│   ├── sample_inputs
│   │   └── model.pt
│   ├── constants
│   │   ├── tensor_0
│   │   ├── tensor_1
│   │   └── model_constants_config.json
│   └── weights
│       ├── weight_0
│       ├── weight_1
│       ├── weight_2
│       ├── weight_3
│       └── model_weights_config.json
└── models
    └── model.json
```

Test Plan:
CI

Rollback Plan:

Differential Revision: D80035490

SherlockNoMad commented on torch/export/unflatten.py:
```diff
@@ -53,6 +53,16 @@ class _AttrKind(Enum):
+    untyped_storage: torch.UntypedStorage
+    stride: tuple
+    size: tuple
+    storage_offset: Union[torch.SymInt, int]
```
If you are apply this parameter and buffers, they should always have concrete offset? 

Not sure how it would end up with symbolic offset?

yiming0416 replied:
@SherlockNoMad Yes, they should always have concrete offset. I make it a Union just to make mypy linter happy. As calling `storage_offset()` on a Tensor could return a `SymInt`

SherlockNoMad commented on torch/csrc/export/pt2_archive_constants.h:
```diff
@@ -33,11 +33,13 @@ namespace torch::_export::archive_spec {
   DO(WEIGHTS_DIR, "data/weights/")                                             \
   DO(WEIGHT_FILENAME_PREFIX, "weight_")                                        \
   DO(WEIGHTS_PARAM_CONFIG_FORMAT, "data/weights/{}_model_param_config.json")   \
+  DO(WEIGHTS_CONFIG_FORMAT, "data/weights/{}_weights_config.json")             \
```
Let's make the name more readable： WEIGHTS_CONFIG_FILENAME_FORMAT

SherlockNoMad commented on torch/nativert/ModelRunner.cpp:
```diff
@@ -114,6 +88,45 @@ ModelRunner::ModelRunner(
+std::unordered_map<std::string, std::string> ModelRunner::getPayloadConfig(
+    const std::shared_ptr<caffe2::serialize::PyTorchStreamReader>&
+        pytorchStreamReader,
+    std::string_view configFormat,
```
configFilenameFormat

yiming0416 commented on torch/nativert/graph/TensorMeta.cpp:
```diff
@@ -106,11 +106,14 @@ TensorMeta::TensorMeta(const torch::_export::TensorMeta& tensorMeta)
-  } else {
-    TORCH_CHECK(false, "SymInt not supported yet");
+  } else if (storageOffset.tag() == torch::_export::SymInt::Tag::AS_EXPR) {
+    // TODO: it's still unclear how SymInt shape should be used in runtime
+    // setting the storage offset to 0 for now
+    hasSymbolicShape_ = true;
+    storage_offset_ = 0;
```
@SherlockNoMad Right, we currently don't support symbolic offset. 
Previously we don't serialize storage offset in `TensorMeta`(it is always set to 0 in `torch/_export/serde/serialize.py` ), that's why C++ deserialization never hits the symbolic branch

Now I need this information and serialize storage_offset in `TensorMeta`. For weights and constants, their offsets will always be concrete. But for outputs, it could have symbolic offsets.

I guess I can change this back to throw exception but will have to mark some nativert mocked unit tests to expected failure.

yiming0416 replied:
@SherlockNoMad  On second thought, it's better to keep this change, for the following reasons:
1. The `storage_offset_` in TensorMeta is not being used anywhere.
2. The `storage_offset_` was always 0 previously.
3. Throwing an exception here could make some models that were previously loadable in NativeRT not loadable anymore due to this deserialization issue.

SherlockNoMad commented on torch/_export/serde/serialize.py:
```diff
@@ -256,6 +256,22 @@ def deserialize_device(d: Device) -> torch.device:
 
 
+def deserialize_size(sizes: Sequence[SymInt]) -> tuple[int, ...]:
+    return tuple(sym_int_size.as_int for sym_int_size in sizes)
```
This only handled the case for int. 

assert sym_int_size is int

SherlockNoMad commented on torch/_export/serde/serialize.py:
```diff
@@ -256,6 +256,22 @@ def deserialize_device(d: Device) -> torch.device:
+
+
+def deserialize_stride(strides: Sequence[SymInt]) -> tuple[int, ...]:
+    return tuple(sym_int_size.as_int for sym_int_size in strides)
```
same here

SherlockNoMad commented on torch/_export/serde/serialize.py:
```diff
@@ -256,6 +256,22 @@ def deserialize_device(d: Device) -> torch.device:
+    return _SERIALIZE_TO_TORCH_DTYPE[st]
+
+
+def deserialize_storage_offset(offset: SymInt) -> int:
```
assert is int

yiming0416 commented on torch/_export/serde/schema.py:
```diff
@@ -458,6 +458,27 @@ class ExportedProgram:
+# payload includes params, buffers, tensor constants, and custom objects.
+@dataclass
+class PayloadMeta:
+    # the path of the payload in the archive file, e.g. "weight_0"
```
@SherlockNoMad Based on our offline discussion, we will use just the filename.
For multiple versions of weights in the future, we will create a subfolder under `data/weights/` namely `data/weights/{model_name}-{variant_name}/` and save the weights and config file of variant there. 
But we always use filename in `PayloadMeta`

SherlockNoMad commented on torch/export/pt2_archive/_package.py:
```diff
@@ -485,6 +651,164 @@ class PT2ArchiveContents:
+        ).to(device)
+    else:
+        # cannot call torch.frombuffer() on empty bytes
+        tensor = torch.zeros(size, dtype=dtype).to(device)
```
torch.zeros(size, dtype=dtype, device=device)

SherlockNoMad commented on torch/export/pt2_archive/_package.py:
```diff
@@ -485,6 +651,164 @@ class PT2ArchiveContents:
+    if len(tensor_bytes) != 0:
+        tensor = torch.frombuffer(
+            tensor_bytes, dtype=dtype, requires_grad=tensor_meta.requires_grad
+        ).to(device)
```
I see, you as_strided it latter. 

Add a comment to explain the tensor here is intermediate, and will be re-strided latter.

SherlockNoMad commented on torch/export/pt2_archive/_package.py:
```diff
@@ -485,6 +651,164 @@ class PT2ArchiveContents:
+        return archive_reader.read_bytes(legacy_weights_file)
+
+    else:
+        weights_config = _dict_to_dataclass(
+            schema.PayloadConfig,
+            json.loads(
+                archive_reader.read_string(WEIGHTS_CONFIG_FORMAT.format(model_name))
+            ),
+        )
```
nit, this can be refactor into a deserialize() util func？

SherlockNoMad reviewed (changes requested):
lgtm.
with the comments addressed, should be good to land

yiming0416 committed:
[1/n][export] Refactor PT2 Archive weight saving and loading (#160394)

Summary:

We split the refactoring in two parts for forward compatibility concerns
First, we land the deserialization (loading part)
Then, we land the serialization (saving part)

Save weights and constants as individual files in PT2 archive. Each weight/constant will be saved as raw bytes, unless it is a custom object (TorchBind object) or a non-fake tensor subclass, for these two special cases we still save them using pickle.

The metadata of saved tensors along with the file name will be saved as `PayloadMeta`.
The mapping from FQN to `PayloadMeta` will be saved as `PayloadConfig` under `WEIGHTS_CONFIG_FORMAT` and `CONTANTS_CONFIG_FORMAT`

This changes the serialization in python side when calling `torch.export.save()`.

For deserialization in python `torch.export.load()`, we make it BC-safe by allowing loading legacy format weights/constants.

For deserialization in C++ `torch/nativert/ModelRunner.cpp`, we make this a BC breaking change as currently the OSS ModelRunner API is not being used.

The file structure

```
├── archive_format
├── archive_version
├── byteorder
├── .data
│   ├── serialization_id
│   └── version
├── data
│   ├── sample_inputs
│   │   └── model.pt
│   ├── constants
│   │   ├── tensor_0
│   │   ├── tensor_1
│   │   └── model_constants_config.json
│   └── weights
│       ├── weight_0
│       ├── weight_1
│       ├── weight_2
│       ├── weight_3
│       └── model_weights_config.json
└── models
    └── model.json
```

Test Plan:
CI

Rollback Plan:

Reviewed By: SherlockNoMad

Differential Revision: D80035490
